libcalico-go/lib/set/adaptive.go

```diff
@@ -0,0 +1,351 @@
+		for i, t := range tSlice {
+			if t == item {
+				// Found the element to remove.
+				newSize := a.size - 1
+				newCap := arrCapForSize[newSize]
+				if newCap < cap(tSlice) {
+					// Downgrade to a smaller array.
+					updatedSlice := make([]T, arrCapForSize[newSize])
+					// Copy the elements before and after the removed element.
+					copy(updatedSlice, tSlice[:i])
+					copy(updatedSlice[i:], tSlice[i+1:])
+					a.p = unsafe.Pointer(&updatedSlice[0])
+				} else {
+					// Keep the same slice.  Swap the last element into the
+					// removed element's slot.
+					tSlice[i] = tSlice[newSize]
+					var zeroT T
+					tSlice[newSize] = zeroT
+				}
+				a.size = newSize
+				return
+			}
+		}
+	case sizeStoredInMap:
+		m := a.loadMapFromPointer()
+		delete(m, item)
+		if len(m) <= adaptiveSetArrayLimit {
+			// Too few items for a map, downgrade to an array.
+			s := make([]T, 0, arrCapForSize[len(m)])
+			for t := range m {
+				s = append(s, t)
+			}
+			a.p = unsafe.Pointer(&s[0])
+			a.size = len(m)
+		}
+	}
+}
+
+func (a *Adaptive[T]) Clear() {
+	a.size = 0
+	a.p = nil
+}
+
+func (a *Adaptive[T]) Contains(t T) bool {
+	switch a.size {
+	case 0:
+		return false
+	default:
+		// Handles sizes 1, 2, ..., adaptiveSetArrayLimit. Stored in an array.
+
+		// Make slice pointing at backing array with correct length/capacity.
+		tSlice := a.loadSliceFromPointer()
+
+		// Scan for the item.
+		for _, v := range tSlice {
+			if v == t {
+				return true
+			}
+		}
+		return false
+	case sizeStoredInMap:
+		m := a.loadMapFromPointer()
+		_, present := m[t]
+		return present
+	}
+}
+
+func (a *Adaptive[T]) Iter(f func(item T) error) {
+	switch a.size {
+	case 0:
+		return
+	default:
+		// Handles sizes 1, 2, ..., adaptiveSetArrayLimit. Stored in an array.
+
+		// Make slice pointing at backing array with correct length/capacity.
+		tSlice := a.loadSliceFromPointer()
+
+		// Take a copy of the whole set so that a.Discard() does not break
+		// iteration.  Since size <= adaptiveSetArrayLimit, we can use a
+		// fixed-size array, which will be stack allocated.
+		var tCopy [adaptiveSetArrayLimit]T
+		copy(tCopy[:], tSlice)
+		tSlice = tCopy[:a.size]
+
+		for _, v := range tSlice {
+			err := f(v)
+			if err == StopIteration {
+				return
+			}
+			if err == RemoveItem {
+				a.Discard(v)
+			}
+		}
+	case sizeStoredInMap:
+		m := a.loadMapFromPointer()
+		for v := range m {
+			err := f(v)
+			if err == StopIteration {
+				return
+			}
+			if err == RemoveItem {
+				// Discarding from a map is safe.  If the set did shrink and
+				// get turned into an array then we'd just keep iterating
+				// over the map, which would be fine.
+				a.Discard(v)
+			}
+		}
+	}
+}
+
+// loadSliceFromPointer loads the array stored in our unsafe pointer, returning
+// it as a slice with len == a.size and cap == arrCapForSize[a.size].
+func (a *Adaptive[T]) loadSliceFromPointer() []T {
+	if a.size < 1 || a.size > adaptiveSetArrayLimit {
```

Comment:
I think `a.size < 1` could be `a.size < 2` if we had the code for downgrading storage from size 2 to size 1.

(Although now wondering if I'm misunderstanding the equivalence between a single value and a slice of len 1.)

Reply:
A pointer to an array is just a pointer to the first element, so, if there's only one element, there's no difference between a pointer to an array and pointer to an element directly.